packages/back-end/src/jobs/refreshFactTableColumns.ts

```diff
@@ -138,7 +138,48 @@ export async function runRefreshColumnsQuery(
 
   const typeMap = new Map<string, FactTableColumnType>();
   const jsonMap = new Map<string, JSONColumnFields>();
-  determineColumnTypes(result.results).forEach((col) => {
+
+  result.columns?.forEach((col) => {
+    // If the underlying SQL engine returned the datatype, use it
+    if (col.dataType !== undefined) {
+      // For JSON, only return if we have the field information, otherwise skip
+      // so we can infer from the returned data
+      if (
+        col.dataType === "json" &&
+        col.fields !== undefined &&
+        col.fields.length > 0
+      ) {
+        typeMap.set(col.name, "json");
+        jsonMap.set(
+          col.name,
+          col.fields.reduce(
+            (acc, field) => ({
+              ...acc,
+              [field.name]: {
+                datatype: field.dataType,
+              },
+            }),
+            {},
+          ),
+        );
+      } else if (col.dataType !== "json") {
+        typeMap.set(col.name, col.dataType);
+      }
+    }
+  });
+
+  // Remove columns that we already know the type of from the first result
+  // so we can determine the types of the remaining columns
+  const resultWithoutKnownColumns = result.results?.[0];
+  if (resultWithoutKnownColumns) {
+    for (const col of Object.keys(resultWithoutKnownColumns)) {
+      if (typeMap.has(col)) {
+        delete resultWithoutKnownColumns[col];
+      }
+    }
+  }
+
+  determineColumnTypes([resultWithoutKnownColumns]).forEach((col) => {
```

Comment:
I think this is a regression. Before, we would pass in all result rows, but now we are only passing in the first. That means in the first row has a null value, this will fail to infer whereas it previously would get the first non-null value from all the rows.

Reply:
Good catch! I missed that we iterate over all `rows` after getting the first one in `determineColumnTypes`.

Fixed.